src/main/scala/io/iohk/ethereum/network/PeerActor.scala

```diff
@@ -270,23 +270,46 @@ class PeerActor(
 
       case RLPxConnectionHandler.MessageReceived(message) =>
         log.debug("Received message: {}", message)
+        updateMaxBlock(message)
         notifySubscribers(message)
         processMessage(message)
 
       case s: SendMessage[_] =>
+        updateMaxBlock(s.message)
         rlpxConnection.sendMessage(s.message)(s.enc)
 
+      case GetMaxBlockNumber(actor) => actor ! MaxBlockNumber(currentPeerMaxBlock)
+
       case GetStatus =>
         sender() ! StatusResponse(Handshaked(initialStatus))
     }
 
-    def notifySubscribers(message: Message): Unit = {
+    private def updateMaxBlock(message: Message) = {
+      message match {
+        case m: BlockHeaders =>
+          update(m.headers.map(_.number))
+        case m: NewBlock =>
+          update(Seq(m.block.header.number))
+        case m: NewBlockHashes =>
+          update(m.hashes.map(_.number))
+        case _ =>
+      }
+
+      def update(ns: Seq[BigInt]) = {
+        val maxBlockNumber = ns.fold(0: BigInt) { case (a, b) => if (a > b) a else b }
+        if (maxBlockNumber > currentPeerMaxBlock) {
+          context become new HandshakedHandler(rlpxConnection, initialStatus, maxBlockNumber).receive
```

Comment:
can we have var for that instead? (maybe pass initialPeerMaxBlock?) if, at some point, someone calls `context.become` before or after `updateMaxBlock` it won't work as expected